src/imex/transfer.rs

```diff
@@ -0,0 +1,539 @@
+
+use crate::blob::BlobDirContents;
+use crate::chat::delete_and_reset_all_device_msgs;
+use crate::context::Context;
+use crate::qr::Qr;
+use crate::{e2ee, EventType};
+
+use super::{export_database, DeleteOnDrop, DBFILE_BACKUP_NAME};
+
+/// Provide or send a backup of this device.
+///
+/// This creates a backup of the current device and starts a service which offers another
+/// device to download this backup.
+///
+/// This does not make a full backup on disk, only the SQLite database is created on disk,
+/// the blobs in the blob directory are not copied.
+///
+/// This starts a task which acquires the global "ongoing" mutex.  If you need to stop the
+/// task use the [`Context::stop_ongoing`] mechanism.
+///
+/// The task implements [`Future`] and awaiting it will complete once a transfer has been
+/// either completed or aborted.
+#[derive(Debug)]
+pub struct BackupProvider {
+    /// The supervisor task, run by [`BackupProvider::watch_provider`].
+    handle: JoinHandle<Result<()>>,
+    /// The ticket to retrieve the backup collection.
+    ticket: Ticket,
+}
+
+impl BackupProvider {
+    /// Prepares for sending a backup to a second device.
+    ///
+    /// Before calling this function all I/O must be stopped so that no changes to the blobs
+    /// or database are happening, this is done by calling the [`Accounts::stop_io`] or
+    /// [`Context::stop_io`] APIs first.
+    ///
+    /// This will acquire the global "ongoing process" mutex, which can be used to cancel
+    /// the process.
+    ///
+    /// [`Accounts::stop_io`]: crate::accounts::Accounts::stop_io
+    pub async fn prepare(context: &Context) -> Result<Self> {
+        e2ee::ensure_secret_key_exists(context)
+            .await
+            .context("Private key not available, aborting backup export")?;
+
+        // Acquire global "ongoing" mutex.
+        let cancel_token = context.alloc_ongoing().await?;
+        let context_dir = context
+            .get_blobdir()
+            .parent()
+            .ok_or(anyhow!("Context dir not found"))?;
+        let dbfile = context_dir.join(DBFILE_BACKUP_NAME);
+        if fs::metadata(&dbfile).await.is_ok() {
+            fs::remove_file(&dbfile).await?;
+            warn!(context, "Previous database export deleted");
+        }
+        let res = tokio::select! {
+            biased;
+            res = Self::prepare_inner(context, &dbfile) => {
+                match res {
+                    Ok(slf) => Ok(slf),
+                    Err(err) => {
+                        error!(context, "Failed to set up second device setup: {:#}", err);
+                        Err(err)
+                    },
+                }
+            },
+            _ = cancel_token.recv() => Err(format_err!("cancelled")),
+        };
+        let (provider, ticket) = match res {
+            Ok((provider, ticket)) => (provider, ticket),
+            Err(err) => {
+                context.free_ongoing().await;
```

Comment:
This has the same async issue, not sure how critical it is here.  Maybe it would be fine to spawn on drop.

However I think we should consider that a separate piece of work: it's really an improvement of the `context.alloc_ongoing()` etc APIs you're proposing.  I agree those are ugly and could be improved, maybe with RAII guards.  But doing so could have some other consequences in other code paths, so it's better to not mingle that into a PR doing other things.